### ISSUE TITLE
fails with "HTTP invalid URI: invalid format" error on some websites

### ISSUE THREAD

**Issue by Shnatsel** [adamreichold/zeptohttpc#6]
On some websites, e.g. http://volvoforums.com, zeptohttpc v0.2.1 fails with the following error:

> HTTP invalid URI: invalid format

Firefox and curl work fine.

1575 websites out of the top million from [Feb 3 Tranco list](https://tranco-list.eu/list/3G6L) are affected.

Tested using [this code](https://github.com/Shnatsel/rust-http-clients-smoke-test/blob/8e3285a45e1d657744a2697ced1bd8461031fb86/zeptohttpc-smoke-test/src/main.rs). Test tool output from all affected websites: [zepto-invalid-uri.tar.gz](https://github.com/adamreichold/zeptohttpc/files/6102827/zepto-invalid-uri.tar.gz)

**Comment by adamreichold** [comment]
This seems to `http::uri::Uri` being unable to parse relative paths with a trailing slash, c.f. https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=79f28833ba458372096f670fd43dfba9

As shown in the playground, `url::Url` does parse this succesfully but not using that dependency was a deliberate decision that I would not like to drop. 

To be honest, I am not verse in the URL spec to determine whether this is actually invalid ATM. But maybe this should be reported against the `http` create?

**Comment by adamreichold** [comment]
It seems that `"forum".parse::<Uri>()` does not parse this as a URI without scheme or authority either:
```rust
[tests/path.rs:5] "forum".parse::<Uri>().unwrap().into_parts() = Parts {
    scheme: None,
    authority: Some(
        forum,
    ),
    path_and_query: None,
    _priv: (),
}
```

**Comment by adamreichold** [comment]
This might be related to https://github.com/hyperium/http/issues/469

**Comment by 89z** [comment]
This still looks to be a problem, at least with these versions:

~~~
http = "0.2.8"
url = "2.2.2"
~~~

contrast with Go:

~~~go
package main
import "net/url"

func main() {
   ref, err := url.Parse("forum/")
   if err != nil {
      panic(err)
   }
   println(ref.Path == "forum/")
}
~~~

which has been able to handle this use case since 2013:

~~~
> .\go.exe version
go version go1.2.2 windows/amd64
> .\go.exe run .\main.go
true
~~~

**Comment by 89z** [comment]
I noticed that `reqwest` is able to handle these, so I checked the code. It seems they do a join, instead of trying to parse the header value directly:

<https://github.com/seanmonstar/reqwest/blob/v0.11.11/src/async_impl/client.rs#L1931-L1936>

This actually matches what Go does as well. It parses the Location in the context of the original URL:

https://github.com/golang/go/blob/go1.19/src/net/http/response.go#L143

**Comment by adamreichold** [comment]
> I noticed that reqwest is able to handle these, so I checked the code. It seems they do a join, instead of trying to parse the header value directly:

Indeed. The problem - if one wants to call it that - is that this would require adding the `url` crate as a dependency here. Or at least reproduce the relevant parts of `Url::join`.

**Comment by 89z** [comment]
I found some other small URI/URL crates that might work as well:

~~~
iref
oni-comb-uri-rs
uriparse
url-parse
~~~

Example:

~~~rust
use iref::{
   Iri,
   IriRefBuf
};

fn main() -> Result<(), iref::Error> {
   let base = Iri::new("https://volvoforums.com")?;
   let resolve = IriRefBuf::new("forum/")?.resolved(base);
   assert!(resolve == "https://volvoforums.com/forum/");
   Ok(())
}
~~~
